### PR TITLE
Refactor abstract tree node

### PR DESCRIPTION
**Pull request by nocommentcode**
Refactored the abstract class `TreeNode`

There was an issue with the previous refactoring. The `TreeNode` class now handles the outer loop of checking all children.
This will make implemeting different Tree Policies even easier

### CODE REVIEW

File: src/main/java/groupM/players/mcts/MCTSPlayer.java [hunk 1]
```diff
@@ -26,7 +26,7 @@ public MCTSPlayer() {
     public MCTSPlayer(long seed) {
```
Comment (grahaminn) [inline src/main/java/groupM/players/mcts/MCTSPlayer.java:26]: Rename here?

File: src/main/java/groupM/players/mcts/MCTSPlayer.java [hunk 2]
```diff
@@ -39,7 +39,7 @@ public MCTSPlayer(long seed) {
     public MCTSPlayer(MCTSParams params) {
```
Comment (grahaminn) [inline src/main/java/groupM/players/mcts/MCTSPlayer.java:39]: Ditto

**Review by grahaminn — CHANGES_REQUESTED** [review]
Noted additional place to rename

File: src/main/java/groupM/players/mcts/TreeNode.java [hunk 3]
```diff
@@ -212,25 +257,32 @@ private boolean finishRollout(AbstractGameState rollerState, int depth) {
...
+     double bestValue = -Double.MAX_VALUE;
 
-    /**
-     * Back up the value of the child through all parents.
-     *
-     * @param result - value of rollout to backup
-     */
-    abstract void backUp(double result);
+     for (AbstractAction action : children.keySet()) {
+        UCB1TreeNode child = (UCB1TreeNode) children.get(action);
```
Comment (grahaminn) [inline src/main/java/groupM/players/mcts/TreeNode.java:265]: What's the relationship between TreeNode and UCB1TreeNode? Should TreeNode assume it's children are of type UCBTreeNode?

**Review by pranigopu — APPROVED** [review]

**Review by grahaminn — APPROVED** [review]
lgtm

**Pull request merged**